plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2600,12 +2757,196 @@ void producer_plugin::log_failed_transaction(const transaction_id_type& trx_id,
+      priority::high,
+      exec_queue::read_only_trx_safe, // placed in read_only_trx_safe queue so it is ensured to be executed in either window
+      [weak_this = weak_from_this()]( const boost::system::error_code& ec ) {
+         auto self = weak_this.lock();
+         if( self && ec != boost::asio::error::operation_aborted ) {
+            self->switch_to_read_window();
+         }
+      }));
+}
+
+// Called from app thread
+void producer_plugin_impl::switch_to_read_window() {
+   EOS_ASSERT(app().executor().is_write_window(),  producer_exception, "expected to be in write window");
+   EOS_ASSERT(_ro_num_active_trx_exec_tasks.load() == 0 && _ro_trx_exec_tasks_fut.empty(),  producer_exception, "_ro_trx_exec_tasks_fut expected to be empty" );
+
+   _ro_write_window_timer.cancel();
+   _ro_read_window_timer.cancel();
+   _time_tracker.add_idle_time( fc::time_point::now() - _idle_trx_time );
+
+   // we are in write window, so no read-only trx threads are processing transactions.
+   // _ro_trx_queue is not being accessed. No need to lock.
+   if ( _ro_trx_queue.queue.empty() ) { // no read-only trxs to process. stay in write window
+      start_write_window(); // restart write window timer for next round
+      return;
+   }
+
+   app().executor().set_to_read_window();
+   chain_plug->chain().set_db_read_only_mode();
+   _received_block = false;
+
+   // start a read-only transaction execution task in each thread in the thread pool
+   _ro_num_active_trx_exec_tasks = _ro_thread_pool_size;
+   for (auto i = 0; i < _ro_thread_pool_size; ++i ) {
+      _ro_trx_exec_tasks_fut.emplace_back( post_async_task( _ro_thread_pool.get_executor(), [self = this] () {
+         return self->read_only_trx_execution_task();
+      }) );
+   }
+
+   auto expire_time = boost::posix_time::microseconds(_ro_read_window_time_us.count());
+   _ro_read_window_timer.expires_from_now( expire_time );
+   _ro_read_window_timer.async_wait( app().executor().wrap(  // stay on app thread
+      priority::high,
+      exec_queue::read_only_trx_safe,
+      [weak_this = weak_from_this()]( const boost::system::error_code& ec ) {
+         auto self = weak_this.lock();
+         if( self && ec != boost::asio::error::operation_aborted ) {
+            // use future to make sure all read-only tasks finished before switching to write window
+            for ( auto& task: self->_ro_trx_exec_tasks_fut ) {
+               task.get();
+            }
+            self->_ro_trx_exec_tasks_fut.clear();
+            self->switch_to_write_window();
+          } else if ( self ) {
+             self->_ro_trx_exec_tasks_fut.clear();
+          }
+       }));
+}
+
+// Called from a read only trx thread. Run in parallel with app and other read only trx threads
+bool producer_plugin_impl::read_only_trx_execution_task() {
+   auto start = fc::time_point::now();
+   auto read_window_deadline = start + _ro_read_window_effective_time_us;
```

Comment:
Pass in `read_window_deadline` so that it is calculated before these are spawning.